docs_src/design/ucr/Microservice-Authentication.md

```diff
@@ -0,0 +1,145 @@
+  revoked, and audited for authorized devices, users, and processes._
+
+
+### Target Users
+- Device Owner
+- Device User
+- Device Maintainer
+- Service Provider
+
+### Description
+
+Microservice authentication provides the following benefits,
+which are potentially valuable to all of the listed target users:
+
+- Provides a defense against malware running on the device,
+  as currently there is no mechanism to ensure that only
+  authorized users or processes are allowed to invoke EdgeX services.
+
+- Provides greater auditability as to who initiated a particular
+  action on the device.
+
+- Depending on implementation, may provide a way to revoke
+  access that was previously granted,
+  or allow customers to tie in to enterprise
+  identity management systems.
+
+For purposes of this UCR, microservice authentication implies
+that the receiving microservice has access to the identity
+of the caller and can write program logic based on that identity.
+
+
+### Existing solutions
+
+Microservice authentication is currently implemented around two primary vectors:
+
+- Token-based authentication schemes.
+  Initiator sends an identifier along with a request to the receiver.
+  The identifier is cryptographically validated using a key trusted by
+  the receiver, or the receiver asks a trusted third party to verify the identifier.
+
+  A benefit of token-based authentication schemes is identity delegation,
+  whereby the identifier can be passed through a chain of calls to
+  preserve the identity of the original initiator.
+  The identifier can often be tunneled through other protocols.
+  Another benefit of token-based authentication is that
+  it flows easily through a web application firewall.
+
+  A drawback of token-based authentication is that due to MITM threats,
+  token-based authentication over an unencrypted network is insecure.
+  Another drawback of token-based authentication is that it is unidirectional:
+  the receiver can authenticate the initiator, but not vice-versa.
+
+- End-to-end encryption schemes.
+  Both the initiator and the receiver participate in a session-oriented message exchange over an encrypted transport,
+  where both parties cryptographically validate each other's identity.
+  
+  A benefit of end-to-end encryption schemes is that it enforces proof-of-possession
+  (of a cryptographic key), thus preventing token-stealing attacks.
+
+  End-to-end encryption schemes are strongest when implemented at the application level,
+  as it does not require trust in the underlying network,
+  with the potential drawback that if network-level encryption is used,
+  or if all communication is done via local IPC interfaces,
+  the system may waste processing power on redundant encryption.
+  Additionally, authentication schemes based on end-to-end encryption
+  may complicate debugging because removing encryption also removes the authentication.
+
+  Service meshes are one example of an end-to-end encryption scheme.
```

Comment:
I don't know all there is to know about service-mesh, but to me this has an assumption that there is an agent/process that runs 'locally' on the machine that the initiator sends traffic to in order to be proxied. The end-to-end encryption provided in that scenario is potentially vulnerable on the local machine. that might be considered a downside. mTLS listed below would not contain that particular vulnerability. that might be worthwhile to call out here?

I would also think that e2ee interferes with reverse proxies as well. That means to me that the callout of mTLS impacting the use of network devices like reverse proxies isn't a problem unique to mTLS, is it?

Reply:
See next update.